examples/src/bin/unbuffered-client.rs

```diff
@@ -0,0 +1,281 @@
+
+// local server
+// const CERTFILE: Option<&str> = Some(concat!(env!("HOME"), "/.local/share/mkcert/rootCA.pem")); // see `mkcert`
+// const SERVER_NAME: &str = "localhost";
+// const PORT: u16 = 1443;
+
+const KB: usize = 1024;
+const INCOMING_TLS_BUFSIZ: usize = 16 * KB;
+const OUTGOING_TLS_INITIAL_BUFSIZ: usize = KB;
+
+const MAX_ITERATIONS: usize = 20;
+
+fn main() -> Result<(), Box<dyn Error>> {
+    // let config = ClientConfig::builder_with_protocol_versions(&[&TLS12])
+    let config = ClientConfig::builder_with_protocol_versions(&[&TLS13])
+        .with_root_certificates(build_root_store()?)
+        .with_no_client_auth();
+
+    let config = Arc::new(config);
+
+    let mut incoming_tls = [0; INCOMING_TLS_BUFSIZ];
+    let mut outgoing_tls = vec![0; OUTGOING_TLS_INITIAL_BUFSIZ];
+
+    converse(&config, false, &mut incoming_tls, &mut outgoing_tls)?;
+
+    Ok(())
+}
+
+fn converse(
+    config: &Arc<ClientConfig>,
+    send_early_data: bool,
+    incoming_tls: &mut [u8],
+    outgoing_tls: &mut Vec<u8>,
+) -> Result<(), Box<dyn Error>> {
+    let mut conn = UnbufferedClientConnection::new(Arc::clone(config), SERVER_NAME.try_into()?)?;
+    let mut sock = TcpStream::connect(format!("{SERVER_NAME}:{PORT}"))?;
+
+    let mut incoming_used = 0;
+    let mut outgoing_used = 0;
+
+    let mut open_connection = true;
+    let mut sent_request = false;
+    let mut received_response = false;
+    let sent_early_data = false;
+
+    let mut iter_count = 0;
+    while open_connection {
+        let UnbufferedStatus { mut discard, state } =
+            conn.process_tls_records(&mut incoming_tls[..incoming_used])?;
+
+        match dbg!(state) {
+            ConnectionState::AppDataAvailable(mut state) => {
+                while let Some(res) = state.next_record() {
+                    let AppDataRecord {
+                        discard: new_discard,
+                        payload,
+                    } = res?;
+                    discard += new_discard;
+
+                    if payload.starts_with(b"HTTP") {
+                        let response = core::str::from_utf8(payload)?;
+                        let header = response
+                            .lines()
+                            .next()
+                            .unwrap_or(response);
+
+                        println!("{header}");
+                    } else {
+                        println!("(.. continued HTTP response ..)");
+                    }
+
+                    received_response = true;
+                }
+            }
+
+            ConnectionState::MustEncodeTlsData(mut state) => {
+                try_or_resize_and_retry(
+                    |out_buffer| state.encode(out_buffer),
+                    |e| {
+                        if let EncodeError::InsufficientSize(is) = &e {
+                            Ok(*is)
+                        } else {
+                            Err(e.into())
+                        }
+                    },
```

Comment:
The `map_err` closure provided to `try_or_resize_and_retry` here and on L143 looks identical. Does this need to be a param? If yes, maybe this common code could be lifted into a helper fn to provide in each case.

Reply:
It is almost identical but it isn't. One of them takes an `EncryptError` and the other one takes an `EncodeError`. I could factor the common bits and replace this `map_err` param with one that tries to transform the error into an `InsufficientSize` value.

Reply:
It is almost identical but it isn't. One of them takes an `EncodeError` and the other takes an `EncryptError`.

Reply:
Ah ok! My mistake.